_episodes/02-cupy.md

```diff
@@ -236,5 +236,103 @@ The linear convolution is actually performed on the GPU, which is shown by a nic
+
+The code we want to run is straightforward. Let us compute all prime numbers between 1 and 10000 on the CPU and see if we can speed it up.
+This is code that you can find on many websites. Small variations are possible, but it will look something like this:
+
+~~~python
+def check_prime_cpu(upper):
+    all_prime_numbers=[]
+    for num in range(2, upper):
+        # all prime numbers are greater than 1
+        for i in range(2, num):
+            if (num % i) == 0:
+                break
+        else:
+            all_prime_numbers.append(num)
+    return all_prime_numbers
+~~~
+{: .source}
+
+Calling "check_prime_cpu(10000)" will return all prime numbers between 1 and 10000 as a list. Let us time it:
+
+~~~python
+%timeit check_prime_cpu(10000) 
+~~~
+{: .source}
+
+You will probably find that "check_prime_cpu" takes several hundreds of milliseconds to complete:
+
+~~~
+378 ms ± 45.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+~~~
+{: .output}
+
+As a quick sidestep, add Numba's JIT (Just in Time compilation) decorator to the "check_prime_cpu" function. You can either add it to the function definition or to the call, so either in this way:
+
+~~~python
+@jit(nopython=True)
+def check_prime_cpu(upper):
+    all_prime_numbers=[]
+    ....
+    ....
+~~~
+{: .source}
+
+or in this way:
+
+~~~python
+%timeit jit(nopython=True)(check_prime_cpu)(upper_limit)
+~~~
+{: .source}
+
+which can give you a timing result similar to this:
+
+~~~
+165 ms ± 19 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+~~~
+{: .output}
+
+So twice as fast, by using a simple decorator. The speedup is much larger for upper = 100000, but that takes a little too much waiting time for this course. Despite the "jit(nopython=True)" decorator the computation is still performed on the CPU. Let us move the computation to the GPU. There are a number of ways to achieve this, one of them is the usage of the "jit(device=True)" decorator, but it depends very much on the nature of the computation. For this example of computing primes we can best use the "vectorize" decorator for a "check_prime_gpu" function that takes an array as input instead of "upper". This is the array we have to use as input for our new "check_prime_gpu" function, instead of upper, a single integer:
+
+~~~python
+np.arange(2, 10000, dtype=np.int32)
+~~~
+{: .source}
+
+So that input to the new "check_prime_gpu" function is simply the array of numbers we need to check for primes. "check_prime_gpu" looks similar to "check_prime_cpu":
```

Comment:
Why are we doing this? Better to at least mention that the structure of CPU and GPU code is different.

Reply:
Okay......you mean why are we not appending to a list when computing on the GPU? 
So I should explain that the outer loop has been removed in order to use the `vectorize` decorator with an array as input instead of a scalar.

Is that what you mean?

Reply:
Yes, you would not normally create such a large array for the output on the CPU, but you are doing it because otherwise you would need to have some dynamic data structure on the GPU and that is much harder to implement.

Reply:
I added an intermediate step with the launch of a CUDA kernel using a cuda.jit decorator. That kernel computes if a number is a prime or not, on the GPU. That intermediate step shows that this is inefficient use of the GPU, because checking if a single number is a prime imposes insufficient workload on the GPU.
The vectorize decorator is then presented as the solution for sufficient workload, because it enables checking of an array of numbers for primes.